sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/PercentileApprox.scala

```diff
@@ -0,0 +1,462 @@
+     memory; higher values yield better approximations.
+    _FUNC_(col, array(p1 [, p2]...) [, B]) - Same as above, but accepts and returns an array of
+     percentile values instead of a single one.
+    """)
+case class PercentileApprox(
+    child: Expression,
+    percentilesExpr: Expression,
+    bExpr: Option[Expression],
+    percentiles: Seq[Double],  // the extracted percentiles
+    B: Int,                    // the extracted B
+    resultAsArray: Boolean,    // whether to return the result as an array
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0) extends ImperativeAggregate {
+
+  private def this(child: Expression, percentilesExpr: Expression, bExpr: Option[Expression]) = {
+    this(
+      child = child,
+      percentilesExpr = percentilesExpr,
+      bExpr = bExpr,
+      // validate and extract percentiles
+      percentiles = PercentileApprox.validatePercentilesLiteral(percentilesExpr)._1,
+      // validate and extract B
+      B = bExpr.map(PercentileApprox.validateBLiteral(_)).getOrElse(PercentileApprox.B_DEFAULT),
+      // validate and mark whether we should return results as array of double or not
+      resultAsArray = PercentileApprox.validatePercentilesLiteral(percentilesExpr)._2)
+  }
+
+  // Constructor for the "_FUNC_(col, p) / _FUNC_(col, array(p1, ...))" form
+  def this(child: Expression, percentilesExpr: Expression) = {
+    this(child, percentilesExpr, None)
+  }
+
+  // Constructor for the "_FUNC_(col, p, B) / _FUNC_(col, array(p1, ...), B)" form
+  def this(child: Expression, percentilesExpr: Expression, bExpr: Expression) = {
+    this(child, percentilesExpr, Some(bExpr))
+  }
+
+  override def prettyName: String = "percentile_approx"
+
+  override def withNewMutableAggBufferOffset(newMutableAggBufferOffset: Int): ImperativeAggregate =
+    copy(mutableAggBufferOffset = newMutableAggBufferOffset)
+
+  override def withNewInputAggBufferOffset(newInputAggBufferOffset: Int): ImperativeAggregate =
+    copy(inputAggBufferOffset = newInputAggBufferOffset)
+
+  override def children: Seq[Expression] =
+    bExpr.map(child :: percentilesExpr :: _ :: Nil).getOrElse(child :: percentilesExpr :: Nil)
+
+  // we would return null for empty inputs
+  override def nullable: Boolean = true
+
+  override def dataType: DataType = if (resultAsArray) ArrayType(DoubleType) else DoubleType
+
+  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, AnyDataType, IntegralType)
+
+  override def checkInputDataTypes(): TypeCheckResult =
+    TypeUtils.checkForNumericExpr(child.dataType, "function percentile_approx")
+
+  // The number of intermediate outputs is highly relative to the actual data-set (an upper bound is
+  // (11/2e)log(2en), where e is the relativeError parameter, n is the number of items in the
+  // dataset) -- thus it's hard to allocate agg buffer in advance without knowing the size of
+  // inputs. Due to this reason, currently we don't support partial mode.
```

Comment:
> hive supports partial aggregate for percentile_approx

Hive's implementation computes approximate percentile values from a histogram, thus Hive supports partial aggregation but makes no approximation guarantees.

> ... QuantileSummaries in this aggregate function object, instead of letting aggregate operator manage it, that's the main reason why we can't support partial aggregate

Yes that's quite right. `QuantileSummaries` has been implemented and well tested prior to this patch, so it'd be great if we can directly put a `QuantileSummaries` instance into the aggregation buffer (in order to support partial aggregation). @cloud-fan any pointer on how to do that?

Thanks!